File: clients/java/cldf-java/src/main/java/io/cldf/models/Location.java
package io.cldf.models;

import java.time.OffsetDateTime;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import io.cldf.models.enums.RockType;
import io.cldf.models.enums.TerrainType;
import io.cldf.models.media.Media;
import io.cldf.utils.FlexibleDateTimeDeserializer;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/** Represents a climbing location such as a crag, boulder area, or climbing gym. */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Location {

  @JsonProperty(required = true)
  private Integer id;

  @JsonProperty(required = true)
  private String name;

  @JsonProperty(required = true)
  private Boolean isIndoor;

  private Coordinates coordinates;

  private String country;

  private String state;

  private String city;

  private String address;

  @Builder.Default private boolean starred = false;

  private RockType rockType;

  private TerrainType terrainType;

  private String accessInfo;

  /** Media associated with this location (overview, approach photos, etc.) */
  private Media media;

  @JsonDeserialize(using = FlexibleDateTimeDeserializer.class)
  private OffsetDateTime createdAt;

  private Map<String, Object> customFields;

  /** Geographic coordinates for a location. */
  @Data
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  public static class Coordinates {
    @JsonProperty(required = true)
    private Double latitude;

    @JsonProperty(required = true)
    private Double longitude;
  }
}

File: clients/dart/cldf/lib/clid/models/location.dart
import 'coordinates.dart';

/// Represents a climbing location for CLID generation
class Location {
  final String country;
  final String? state;
  final String? city;
  final String name;
  final Coordinates coordinates;
  final bool isIndoor;

  const Location({
    required this.country,
    this.state,
    this.city,
    required this.name,
    required this.coordinates,
    required this.isIndoor,
  });

  factory Location.outdoor({
    required String country,
    String? state,
    String? city,
    required String name,
    required double lat,
    required double lon,
  }) {
    return Location(
      country: country,
      state: state,
      city: city,
      name: name,
      coordinates: Coordinates(lat: lat, lon: lon),
      isIndoor: false,
    );
  }

  factory Location.indoor({
    required String country,
    String? state,
    String? city,
    required String name,
    required double lat,
    required double lon,
  }) {
    return Location(
      country: country,
      state: state,
      city: city,
      name: name,
      coordinates: Coordinates(lat: lat, lon: lon),
      isIndoor: true,
    );
  }

  Map<String, dynamic> toJson() => {
    'country': country,
    if (state != null) 'state': state,
    if (city != null) 'city': city,
    'name': name,
    'coordinates': coordinates.toJson(),
    'isIndoor': isIndoor,
  };

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      country: json['country'] as String,
      state: json['state'] as String?,
      city: json['city'] as String?,
      name: json['name'] as String,
      coordinates: Coordinates.fromJson(
        json['coordinates'] as Map<String, dynamic>,
      ),
      isIndoor: json['isIndoor'] as bool,
    );
  }
}
